src/java.base/share/classes/sun/security/provider/ByteArrayAccess.java

```diff
@@ -59,439 +49,149 @@ private ByteArrayAccess() {
-            out[ 2] = reverseBytes(unsafe.getInt(in, (long)(inOfs +  8)));
-            out[ 3] = reverseBytes(unsafe.getInt(in, (long)(inOfs + 12)));
-            out[ 4] = reverseBytes(unsafe.getInt(in, (long)(inOfs + 16)));
-            out[ 5] = reverseBytes(unsafe.getInt(in, (long)(inOfs + 20)));
-            out[ 6] = reverseBytes(unsafe.getInt(in, (long)(inOfs + 24)));
-            out[ 7] = reverseBytes(unsafe.getInt(in, (long)(inOfs + 28)));
-            out[ 8] = reverseBytes(unsafe.getInt(in, (long)(inOfs + 32)));
-            out[ 9] = reverseBytes(unsafe.getInt(in, (long)(inOfs + 36)));
-            out[10] = reverseBytes(unsafe.getInt(in, (long)(inOfs + 40)));
-            out[11] = reverseBytes(unsafe.getInt(in, (long)(inOfs + 44)));
-            out[12] = reverseBytes(unsafe.getInt(in, (long)(inOfs + 48)));
-            out[13] = reverseBytes(unsafe.getInt(in, (long)(inOfs + 52)));
-            out[14] = reverseBytes(unsafe.getInt(in, (long)(inOfs + 56)));
-            out[15] = reverseBytes(unsafe.getInt(in, (long)(inOfs + 60)));
-        } else {
-            b2iLittle(in, inOfs, out, 0, 64);
+        len += inOfs;
+        while (inOfs < len) {
+            out[outOfs++] = (int) LE.INT_ARRAY.get(in, inOfs);
+            inOfs += 4;
         }
     }
 
     /**
      * int[] to byte[] conversion, little endian byte order.
      */
     static void i2bLittle(int[] in, int inOfs, byte[] out, int outOfs, int len) {
-        if ((inOfs < 0) || ((in.length - inOfs) < len/4) ||
-            (outOfs < 0) || ((out.length - outOfs) < len)) {
-            throw new ArrayIndexOutOfBoundsException();
-        }
-        if (littleEndianUnaligned) {
-            outOfs += byteArrayOfs;
-            len += outOfs;
-            while (outOfs < len) {
-                unsafe.putInt(out, (long)outOfs, in[inOfs++]);
-                outOfs += 4;
-            }
-        } else if (bigEndian && ((outOfs & 3) == 0)) {
-            outOfs += byteArrayOfs;
-            len += outOfs;
-            while (outOfs < len) {
-                unsafe.putInt(out, (long)outOfs, reverseBytes(in[inOfs++]));
-                outOfs += 4;
-            }
-        } else {
-            len += outOfs;
-            while (outOfs < len) {
-                int i = in[inOfs++];
-                out[outOfs++] = (byte)(i      );
-                out[outOfs++] = (byte)(i >>  8);
-                out[outOfs++] = (byte)(i >> 16);
-                out[outOfs++] = (byte)(i >> 24);
-            }
+        len += outOfs;
+        while (outOfs < len) {
+            LE.INT_ARRAY.set(out, outOfs, in[inOfs++]);
+            outOfs += 4;
         }
     }
 
     // Store one 32-bit value into out[outOfs..outOfs+3] in little endian order.
     static void i2bLittle4(int val, byte[] out, int outOfs) {
-        if ((outOfs < 0) || ((out.length - outOfs) < 4)) {
-            throw new ArrayIndexOutOfBoundsException();
-        }
```

Comment:
Why do we remove the index checking from all methods? Isn't it safer to check here in case the caller didn't?

Reply:
Yes, IOOBE checking is done by the VarHandle methods, while the Unsafe API is unsafe and needs careful precondition checking. It doesn't seem to matter for performance (interpreted code sees some benefit by the removal).

With the current usage an IOOBE is probably not observable, but there's a test that reflects into ByteArrayAccess and verifies exceptions are thrown as expected on faulty inputs.